### ISSUE TITLE
Query starts throwing exception after upgrade to Neo4j 3.0.4

### ISSUE THREAD

**Issue by rationalism** [neo4j/neo4j#7713]
- Neo4j version: 3.0.4
- Operating system: Ubuntu 16.04
- API/Driver: Java driver version 1.0.4, but also confirmed in Neo4j browser 

When I upgraded to Neo4j 3.0.4 from 3.0.3, this query (and many other similar queries) started throwing a DatabaseException:

```
MATCH (root) WHERE ID(root) = 1110642 
WITH root 
OPTIONAL MATCH (root)-[r:`name`]-(a0) 
WITH root, collect([ID(root), labels(root), type(r), a0.`val`, r]) AS b0 
OPTIONAL MATCH (root)-[r:`email-addr`]-(a1) 
WITH root, b0, collect([ID(root), labels(root), type(r), a1.`val`, r]) AS b1 
OPTIONAL MATCH (root)-[r:`phone-num`]-(a2) WITH root, b0, b1, collect([ID(root), labels(root), type(r), a2.`val`, r]) AS b2 
OPTIONAL MATCH (root)-[:`email-from`]-(b0a3)-[r:`sent-date`]-(a3) 
WITH root, b0, b1, b2, collect([ID(b0a3), labels(b0a3), type(r), a3.`val`, r]) AS b3 
OPTIONAL MATCH (root)-[:`email-to`]-(b0a4)-[r:`sent-date`]-(a4) 
WITH root, b0, b1, b2, b3, collect([ID(b0a4), labels(b0a4), type(r), a4.`val`, r]) AS b4 
OPTIONAL MATCH (root)-[:`link-to`]-(b0a5)-[:`has-link`]-(b1a5)-[r:`sent-date`]-(a5) 
WITH root, b0, b1, b2, b3, b4, collect([ID(b1a5), labels(b1a5), type(r), a5.`val`, r]) AS b5 
RETURN b0, b1, b2, b3, b4, b5
```

When I tried this query in the Neo4j browser, it returned this error message:

(ArrayBuffer(1110642, (recon,person_user_1062118), null),null) (of class scala.Tuple2)

Here is the stack trace in logs/debug.log:

https://gist.github.com/rationalism/347a085fb9bb3ec916d04449c692976a

The problem was fixed by reverting to Neo4j version 3.0.3.

**Comment by rationalism** [comment]
Poking through the changelog, perhaps this is related to issue #7407?

**Comment by jexp** [comment]
Thanks a lot for your feedback!

It's related to the mixed type collections which are not recommended so there was no test to catch this issue. 

We'll look into it. 

Von meinem iPhone gesendet

**Comment by rationalism** [comment]
Thanks! Is there a way to rewrite a query like this:

MATCH (root) WHERE ID(root) = 1110642 WITH root OPTIONAL MATCH (root)-[r:name]-(a0) WITH root, collect([ID(root), labels(root), type(r), a0.val, r]) AS b0 RETURN b0

so that it still returns the same results, but without the mixed type collection? I tried several simpler ways of writing it, but they all returned either too many rows, or too few.

**Comment by jexp** [comment]
@rationalism yes you can use a map instead:

```
MATCH (root) WHERE ID(root) = 1110642 
OPTIONAL MATCH (root)-[r:name]-(a0)
WITH root, collect({ id : ID(root), labels: labels(root), type : type(r), val : a0.val, rel : r} ) AS b0 
RETURN b0
```

you can also simplify it to:

```
MATCH (root) WHERE ID(root) = 1110642 
OPTIONAL MATCH (root)-[r:name]-(a0)
RETURN collect({ id : ID(root), labels: labels(root), type : type(r), val : a0.val, rel : r} ) AS b0 
```

**Comment by rationalism** [comment]
@jexp That worked. Thanks!

**Comment by Mats-SX** [comment]
This is fixed by #7936
